core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -612,17 +612,14 @@ private RelDataType getTightestCommonTypeOrThrow(
     }
 
     if (SqlTypeUtil.isExactNumeric(type1) && SqlTypeUtil.isExactNumeric(type2)) {
-      if (SqlTypeUtil.isDecimal(type1)) {
-        // Use max precision
+      if (SqlTypeUtil.isDecimal(type1) || SqlTypeUtil.isDecimal(type2)) {
+        // Precision used must be large enough to fit either of the types
+        int maxScale = Math.max(type1.getScale(), type2.getScale());
         RelDataType result =
-            factory.createSqlType(type1.getSqlTypeName(),
-                Math.max(type1.getPrecision(), type2.getPrecision()), type1.getScale());
-        return factory.createTypeWithNullability(result, type1.isNullable() || type2.isNullable());
-      } else if (SqlTypeUtil.isDecimal(type2)) {
-        // Use max precision
-        RelDataType result =
-            factory.createSqlType(type2.getSqlTypeName(),
-                Math.max(type1.getPrecision(), type2.getPrecision()), type2.getScale());
+            factory.createSqlType(SqlTypeName.DECIMAL,
+                Math.max(type1.getPrecision() - type1.getScale(),
+                         type2.getPrecision() - type2.getScale()) + maxScale,
+                maxScale);
         return factory.createTypeWithNullability(result, type1.isNullable() || type2.isNullable());
       }
       if (type1.getPrecision() > type2.getPrecision()) {
```

Comment:
Can this handle types without precision and proportion? For example:
TINYINT、SMALLINT、INTEGER and BIGINT? Or the default behavior is right.
If type1 is TINYINT type2 is SMALLINT, then the common type becomes SMALLINT. This is right.
But when type1 is SMALLINT and  type2 is TINYINT. then the common type becomes TINYINT. This should be unreasonable.

Reply:
this should be completely symmetric in type1 and type2.

Reply:
BTW: all integer types have a positive precision and a scale or zero.

Reply:
According to SqlTypeName, TINYINT、SMALLINT、INTEGER, and BIGINT don't have Precision and Scale.

Reply:
See RelDataTypeSystemImpl.getDefaultPrecision

Reply:
But the `RelDataTypeSystemImpl#getDefaultPrecision` has a default precision for this. It is right. I misunderstood.